File: backend/agent/safety/_policy.py
"""Tool-level safety assessment."""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from agent.safety._filesystem import is_dangerous_path, is_path_within_workspace
from agent.safety._powershell import classify as classify_powershell

# Maps each tool name to a permission group.
TOOL_GROUP: dict[str, str] = {
    "bash": "shell",
    "powershell": "shell",
    "file_read": "files",
    "read": "files",
    "file_write": "files",
    "write": "files",
    "file_delete": "files",
    "delete": "files",
    "file_list": "files",
    "list": "files",
    "glob": "files",
    "grep": "files",
    "edit": "files",
    "file_move": "files",
    "file_copy": "files",
    "apply_patch": "files",
    "archive": "files",
    "lsp": "files",
    "google_search": "search",
    "youtube_search": "search",
    "browser_open": "search",
    "browser_read": "search",
    "browser_screenshot": "search",
    "webfetch": "search",
    "http_request": "search",
    "browser_click": "search",
    "browser_fill": "search",
    "mouse": "pc_control",
    "keyboard": "pc_control",
    "screenshot": "pc_control",
    "clipboard": "pc_control",
    "window_list": "pc_control",
    "window_focus": "pc_control",
    "window_manage": "pc_control",
    "image_locate": "pc_control",
    "ocr": "pc_control",
    "wait_for": "pc_control",
    "process_list": "processes",
    "process_kill": "processes",
    "system_info": "system",
    "notify": "system",
}

# Canonical group order for the UI.
PERMISSION_GROUPS = ["files", "shell", "search", "pc_control", "processes", "system"]


@dataclass
class SafetyDecision:
    risk: str  # "safe" | "confirm" | "blocked"
    reason: str
    summary: str
    details: list[str] = field(default_factory=list)
    normalized_args: dict[str, Any] = field(default_factory=dict)


def _decide(
    risk, reason, summary, details=None, args=None, tool=None, mode="ask"
) -> SafetyDecision:
    d = SafetyDecision(
        risk=risk, reason=reason, summary=summary, details=details or [], normalized_args=args or {}
    )
    return _apply_mode(d, tool or "", mode)


def _apply_mode(decision: SafetyDecision, tool_name: str, mode: str) -> SafetyDecision:
    # In auto mode, `confirm` is normally promoted to `safe` so the agent can
    # run without prompting on every step. A few operations stay gated even in
    # auto mode because they can escape the workspace or are destructive:
    #   - file_delete / delete                 — always destructive
    #   - file_write / edit / apply_patch       — when the target path is
    #                                              outside the workspace (the
    #                                              decision reason carries the
    #                                              "outside workspace" marker)
    #   - file_read / file_list                 — outside workspace reads stay
    #                                              gated so secrets (.env,
    #                                              ~/.ssh, browser cookies) are
    #                                              not silently exfiltrated
    if mode == "auto" and decision.risk == "confirm":
        # apply_patch can target arbitrary paths inside the patch text that
        # the safety layer cannot inspect ahead of time, so it always stays
        # gated in auto mode.
        always_gated = tool_name in ("file_delete", "delete", "apply_patch")
        outside_workspace = "outside workspace" in decision.reason
        if not always_gated and not outside_workspace:
            return SafetyDecision(
                risk="safe",
                reason=decision.reason,
                summary=decision.summary,
                details=decision.details,
                normalized_args=decision.normalized_args,
            )
    return decision


def assess_tool_call(
    tool_name: str,
    args: dict,
    cwd: str | None = None,
    mode: str = "ask",
    permissions: dict[str, str] | None = None,
) -> SafetyDecision:
    if cwd is None:
        cwd = os.getcwd()
    workspace = Path(cwd).resolve()
    norm = dict(args)

    # Apply per-group permission overrides before the policy runs.
    if permissions:
        group = TOOL_GROUP.get(tool_name)
        if group:
            perm = permissions.get(group)
            if perm == "block":
                return SafetyDecision(
                    risk="blocked",
                    reason="disabled in permission settings",
                    summary="Blocked in settings",
                    normalized_args=norm,
                )
            elif perm == "allow":
                mode = "auto"
            elif perm == "ask":
                mode = "ask"

    def _d(risk, reason, summary, details=None):
        return _decide(risk, reason, summary, details, norm, tool_name, mode)

    # file_write
    if tool_name in ("file_write", "write"):
        path = str(norm.get("path", ""))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "File path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "writes outside workspace", "Writing file outside workspace")
        return _d("confirm", "modifies files", "Writing file inside workspace")

    # file_delete
    if tool_name in ("file_delete", "delete"):
        path = str(norm.get("path", ""))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "File path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d(
                "blocked", "deletes outside workspace", "Deleting file outside workspace is blocked"
            )
        return _d("confirm", "destructive file operation", "Deleting file inside workspace")

    # file_read
    if tool_name in ("file_read", "read"):
        path = str(norm.get("path", ""))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "File path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "reads outside workspace", "Reading file outside workspace")
        return _d("safe", "read-only", "Reading file")

    # edit (in-place string replacement in a file)
    if tool_name == "edit":
        path = str(norm.get("path", ""))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "File path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "writes outside workspace", "Editing file outside workspace")
        return _d("confirm", "modifies files", "Editing file inside workspace")

    # glob (file discovery by pattern)
    if tool_name == "glob":
        path = str(norm.get("path") or ".")
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "Path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "lists outside workspace", "Globbing outside workspace")
        return _d("safe", "read-only", "Globbing inside workspace")

    # grep (content search)
    if tool_name == "grep":
        path = str(norm.get("path") or ".")
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "Path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "reads outside workspace", "Searching outside workspace")
        return _d("safe", "read-only", "Searching inside workspace")

    # lsp (language server queries — read-only by design)
    if tool_name == "lsp":
        path = str(norm.get("path") or norm.get("root") or ".")
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "Path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "reads outside workspace", "LSP query outside workspace")
        return _d("safe", "read-only", "LSP query inside workspace")

    # file_list
    if tool_name in ("file_list", "list"):
        path = str(norm.get("path", "."))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "Path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(path, workspace):
            return _d("confirm", "lists outside workspace", "Listing directory outside workspace")
        return _d("safe", "read-only", "Listing directory")

    # todowrite / skill
    if tool_name == "todowrite":
        return _d("safe", "updates session todo state", "Updating todo list")
    if tool_name == "skill":
        return _d(
            "safe", "reads local skill files", "Loading skill", [f"name: {norm.get('name', '')}"]
        )

    # bash / powershell
    if tool_name in ("bash", "powershell"):
        command = str(norm.get("command", ""))
        risk, reason, details = classify_powershell(command)
        summary = "Read-only shell command" if risk == "safe" else reason.capitalize()
        return _d(risk, reason, summary, details)

    # clipboard
    if tool_name == "clipboard":
        if str(norm.get("action", "read")).lower() == "read":
            return _d("safe", "read-only", "Reading clipboard")
        return _d("confirm", "modifies clipboard", "Writing to clipboard")

    # screenshot
    if tool_name == "screenshot":
        return _d("safe", "read-only", "Taking screenshot")

    # mouse
    if tool_name == "mouse":
        action = str(norm.get("action", "click")).lower()
        if action == "move":
            return _d("safe", "read-only pointer", "Moving cursor")
        return _d(
            "confirm", "simulates input", f"Mouse {action}", ["can interact with UI elements"]
        )

    # keyboard
    if tool_name == "keyboard":
        action = str(norm.get("action", "type")).lower()
        text = str(norm.get("text", "")).lower()
        if action == "press":
            dangerous = {"delete", "backspace", "alt+f4", "ctrl+w", "ctrl+shift+w"}
            if any(dk in text for dk in dangerous):
                return _d(
                    "confirm",
                    "destructive key combination",
                    f"Pressing {text}",
                    ["can close windows or delete content"],
                )
        return _d("confirm", "simulates input", f"Keyboard {action}", ["types or presses keys"])

    # browser_open
    if tool_name == "browser_open":
        url = str(norm.get("url", "")).lower()
        if "localhost" in url or "127.0.0.1" in url:
            return _d("safe", "local URL", "Opening localhost URL", [])
        return _d("confirm", "opens external URL", "Opening external URL", [])

    # read-only browser / search tools
    if tool_name in ("browser_read", "browser_screenshot", "youtube_search", "google_search"):
        return _d("safe", "read-only", f"Using {tool_name}")

    # apply_patch
    if tool_name == "apply_patch":
        return _d(
            "confirm",
            "modifies files via patch",
            "Applying patch to files",
            ["can create, modify, delete, or rename files"],
        )

    # webfetch
    if tool_name == "webfetch":
        url = str(norm.get("url", "")).lower()
        if "localhost" in url or "127.0.0.1" in url or "::1" in url:
            return _d("safe", "read-only local", "Fetching local URL", [f"url: {url}"])
        return _d("safe", "read-only", f"Fetching {url}")

    # question
    if tool_name == "question":
        return _d("safe", "interactive", "Asking user")

    # process_list / process_kill
    if tool_name == "process_list":
        return _d("safe", "read-only", "Listing processes")
    if tool_name == "process_kill":
        return _d(
            "confirm",
            "terminates a process",
            "Killing process",
            ["can disrupt the system or other applications"],
        )

    # file_move / file_copy
    if tool_name in ("file_move", "file_copy"):
        src = str(norm.get("src", ""))
        dest = str(norm.get("dest", ""))
        if is_dangerous_path(src) or is_dangerous_path(dest):
            return _d(
                "blocked",
                "dangerous path",
                "Path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if not is_path_within_workspace(src, workspace) or not is_path_within_workspace(
            dest, workspace
        ):
            return _d(
                "blocked",
                "path outside workspace",
                "file_move/file_copy outside workspace is blocked",
            )
        return _d("confirm", "mutates filesystem", f"{tool_name} inside workspace")

    # archive
    if tool_name == "archive":
        action = str(norm.get("action", "list")).lower()
        path = str(norm.get("path", ""))
        if is_dangerous_path(path):
            return _d(
                "blocked",
                "dangerous path",
                "Archive path is outside allowed scope",
                ["UNC path, device path, or traversal detected"],
            )
        if action == "list":
            return _d("safe", "read-only", "Listing archive")
        if action == "create":
            sources = norm.get("sources", []) or []
            if not is_path_within_workspace(path, workspace):
                return _d("blocked", "path outside workspace", "Archive path is outside workspace")
            for s in sources:
                if is_dangerous_path(str(s)) or not is_path_within_workspace(str(s), workspace):
                    return _d(
                        "blocked", "source outside workspace", "Archive source is outside workspace"
                    )
            return _d("confirm", "creates archive", "Creating archive")
        if action == "extract":
            dest = str(norm.get("dest", "")) or path
            if is_dangerous_path(dest) or not is_path_within_workspace(dest, workspace):
                return _d("blocked", "path outside workspace", "Extract dest is outside workspace")
            return _d("confirm", "extracts archive", "Extracting archive")
        return _d(
            "confirm",
            "unknown archive action",
            f"archive {action}",
            ["action must be list, extract, or create"],
        )

    # window tools
    if tool_name == "window_list":
        return _d("safe", "read-only", "Listing windows")
    if tool_name == "window_focus":
        return _d(
            "confirm",
            "changes foreground window",
            "Focusing window",
            [f"title: {norm.get('title', '')}", f"hwnd: {norm.get('hwnd', '')}"],
        )
    if tool_name == "window_manage":
        action = str(norm.get("action", "")).lower()
        return _d(
            "confirm",
            "manipulates a window",
            f"Window {action}",
            ["can move, resize, minimize, maximize, or close windows"],
        )

    # screen perception
    if tool_name == "image_locate":
        return _d("safe", "read-only", "Locating image on screen")
    if tool_name == "ocr":
        return _d("safe", "read-only", "Recognizing text on screen")
    if tool_name == "wait_for":
        return _d(
            "safe",
            "read-only polling",
            f"Waiting for {norm.get('type', '')}",
            [f"target: {norm.get('target', '')}"],
        )

    # system_info / notify
    if tool_name == "system_info":
        return _d("safe", "read-only", "Reading system info")
    if tool_name == "notify":
        return _d("safe", "shows a notification", "Sending desktop notification")

    # memory
    if tool_name == "memory":
        return _d("safe", "local notes store", f"memory {norm.get('action', '')}")

    # http_request
    if tool_name == "http_request":
        method = str(norm.get("method", "GET")).upper()
        url = str(norm.get("url", ""))
        if method in ("GET", "HEAD", "OPTIONS"):
            return _d("safe", "read-only request", f"{method} {url}")
        return _d(
            "confirm",
            "sends a write request",
            f"{method} {url}",
            ["can create or modify remote state"],
        )

    # interactive browser
    if tool_name == "browser_click":
        return _d(
            "confirm",
            "interacts with a web page",
            "Clicking page element",
            [f"selector: {norm.get('selector', '')}"],
        )
    if tool_name == "browser_fill":
        return _d(
            "confirm",
            "interacts with a web page",
            "Filling page field",
            [f"selector: {norm.get('selector', '')}"],
        )

    return _d(
        "confirm",
        "unknown tool",
        f"Unknown tool: {tool_name}",
        ["no safety policy defined — requires confirmation"],
    )

File: agent/safety/policy.go
package safety

import (
	"fmt"
	"os"
	"strings"
)

type SafetyDecision struct {
	Risk           string
	Reason         string
	Summary        string
	Details        []string
	NormalizedArgs map[string]any
}

func getString(m map[string]any, key string) string {
	v, _ := m[key].(string)
	return v
}

func getBool(m map[string]any, key string) bool {
	v, _ := m[key].(bool)
	return v
}

func applyMode(d SafetyDecision, toolName string, mode string) SafetyDecision {
	if mode == "auto" && d.Risk == "confirm" && toolName != "file_delete" && toolName != "delete" && toolName != "apply_patch" {
		return SafetyDecision{
			Risk: "safe", Reason: d.Reason, Summary: d.Summary,
			Details: d.Details, NormalizedArgs: d.NormalizedArgs,
		}
	}
	return d
}

func AssessToolCall(toolName string, args map[string]any, cwd string, mode string) SafetyDecision {
	if cwd == "" {
		cwd, _ = os.Getwd()
	}
	workspace := cwd

	norm := make(map[string]any, len(args))
	for k, v := range args {
		norm[k] = v
	}

	dec := func(risk, reason, summary string, details ...string) SafetyDecision {
		return applyMode(SafetyDecision{
			Risk: risk, Reason: reason, Summary: summary,
			Details: details, NormalizedArgs: norm,
		}, toolName, mode)
	}

	switch toolName {
	case "file_write", "write":
		path := getString(norm, "path")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "File path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(path, workspace) {
			return dec("blocked", "writes outside workspace", "Writing file outside workspace is blocked")
		}
		return dec("confirm", "modifies files", "Writing file inside workspace")

	case "file_delete", "delete":
		path := getString(norm, "path")
		recursive := getBool(norm, "recursive")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "File path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(path, workspace) {
			return dec("blocked", "deletes outside workspace", "Deleting file outside workspace is blocked")
		}
		var details []string
		if recursive {
			details = append(details, "recursive directory deletion")
		}
		return dec("confirm", "destructive file operation", "Deleting file inside workspace", details...)

	case "file_read", "read":
		path := getString(norm, "path")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "File path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(path, workspace) {
			return dec("confirm", "reads outside workspace", "Reading file outside workspace")
		}
		return dec("safe", "read-only", "Reading file")

	case "file_list", "list":
		path := getString(norm, "path")
		if path == "" {
			path = "."
		}
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "Path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(path, workspace) {
			return dec("confirm", "lists outside workspace", "Listing directory outside workspace")
		}
		return dec("safe", "read-only", "Listing directory")

	case "todowrite":
		return dec("safe", "updates session todo state", "Updating todo list")

	case "skill":
		return dec("safe", "reads local skill files", "Loading skill",
			fmt.Sprintf("name: %s", getString(norm, "name")))

	case "bash", "powershell":
		command := getString(norm, "command")
		risk, reason, details := classify(command)
		summary := "Read-only shell command"
		if risk != "safe" {
			summary = capitalize(reason)
		}
		return applyMode(SafetyDecision{
			Risk: risk, Reason: reason, Summary: summary,
			Details: details, NormalizedArgs: norm,
		}, toolName, mode)

	case "clipboard":
		if strings.ToLower(getString(norm, "action")) == "read" {
			return dec("safe", "read-only", "Reading clipboard")
		}
		return dec("confirm", "modifies clipboard", "Writing to clipboard")

	case "screenshot", "screenshot_region", "window_screenshot":
		return dec("safe", "read-only", "Taking screenshot")

	case "mouse":
		action := strings.ToLower(getString(norm, "action"))
		if action == "" {
			action = "click"
		}
		if action == "move" {
			return dec("safe", "read-only pointer", "Moving cursor")
		}
		return dec("confirm", "simulates input", fmt.Sprintf("Mouse %s", action),
			"can interact with UI elements")

	case "keyboard":
		action := strings.ToLower(getString(norm, "action"))
		if action == "" {
			action = "type"
		}
		text := strings.ToLower(getString(norm, "text"))
		if action == "press" {
			dangerous := map[string]bool{
				"delete": true, "backspace": true, "alt+f4": true,
				"ctrl+w": true, "ctrl+shift+w": true,
			}
			for dk := range dangerous {
				if strings.Contains(text, dk) {
					return dec("confirm", "destructive key combination",
						fmt.Sprintf("Pressing %s", text),
						"can close windows or delete content")
				}
			}
		}
		return dec("confirm", "simulates input", fmt.Sprintf("Keyboard %s", action),
			"types or presses keys")

	case "browser_open":
		url := strings.ToLower(getString(norm, "url"))
		if strings.Contains(url, "localhost") || strings.Contains(url, "127.0.0.1") {
			return dec("safe", "local URL", "Opening localhost URL",
				fmt.Sprintf("url: %s", url))
		}
		return dec("confirm", "opens external URL", "Opening external URL",
			fmt.Sprintf("url: %s", url))

	case "browser_read", "browser_screenshot", "youtube_search", "google_search":
		return dec("safe", "read-only", fmt.Sprintf("Using %s", toolName))

	case "apply_patch":
		return dec("confirm", "modifies files via patch", "Applying patch to files",
			"can create, modify, delete, or rename files")

	case "webfetch":
		url := strings.ToLower(getString(norm, "url"))
		if strings.Contains(url, "localhost") || strings.Contains(url, "127.0.0.1") || strings.Contains(url, "::1") {
			return dec("safe", "read-only local", "Fetching local URL",
				fmt.Sprintf("url: %s", url))
		}
		return dec("safe", "read-only", fmt.Sprintf("Fetching %s", url))

	case "question":
		return dec("safe", "interactive", "Asking user")

	case "process_list":
		return dec("safe", "read-only", "Listing processes")

	case "process_kill":
		return dec("confirm", "terminates a process", "Killing process",
			"can disrupt the system or other applications")

	case "file_move", "file_copy":
		src := getString(norm, "src")
		dest := getString(norm, "dest")
		if isDangerousPath(src) || isDangerousPath(dest) {
			return dec("blocked", "dangerous path", "Path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(src, workspace) || !isPathWithinWorkspace(dest, workspace) {
			return dec("blocked", "path outside workspace",
				"file_move/file_copy outside workspace is blocked")
		}
		return dec("confirm", "mutates filesystem", fmt.Sprintf("%s inside workspace", toolName))

	case "archive":
		action := strings.ToLower(getString(norm, "action"))
		if action == "" {
			action = "list"
		}
		path := getString(norm, "path")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "Archive path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		switch action {
		case "list":
			return dec("safe", "read-only", "Listing archive")
		case "create":
			sources, _ := norm["sources"].([]any)
			if !isPathWithinWorkspace(path, workspace) {
				return dec("blocked", "path outside workspace", "Archive path is outside workspace")
			}
			for _, s := range sources {
				sStr := fmt.Sprint(s)
				if isDangerousPath(sStr) || !isPathWithinWorkspace(sStr, workspace) {
					return dec("blocked", "source outside workspace", "Archive source is outside workspace")
				}
			}
			return dec("confirm", "creates archive", "Creating archive")
		case "extract":
			dest := getString(norm, "dest")
			if dest == "" {
				dest = path
			}
			if isDangerousPath(dest) || !isPathWithinWorkspace(dest, workspace) {
				return dec("blocked", "path outside workspace", "Extract dest is outside workspace")
			}
			return dec("confirm", "extracts archive", "Extracting archive")
		default:
			return dec("confirm", "unknown archive action", fmt.Sprintf("archive %s", action),
				"action must be list, extract, or create")
		}

	case "window_list":
		return dec("safe", "read-only", "Listing windows")

	case "window_focus":
		return dec("confirm", "changes foreground window", "Focusing window",
			fmt.Sprintf("title: %s", getString(norm, "title")),
			fmt.Sprintf("hwnd: %s", getString(norm, "hwnd")))

	case "window_manage":
		action := strings.ToLower(getString(norm, "action"))
		return dec("confirm", "manipulates a window", fmt.Sprintf("Window %s", action),
			"can move, resize, minimize, maximize, or close windows")

	case "image_locate":
		return dec("safe", "read-only", "Locating image on screen")

	case "ocr":
		return dec("safe", "read-only", "Recognizing text on screen")

	case "wait_for":
		return dec("safe", "read-only polling",
			fmt.Sprintf("Waiting for %s", getString(norm, "type")),
			fmt.Sprintf("target: %s", getString(norm, "target")))

	case "system_info":
		return dec("safe", "read-only", "Reading system info")

	case "now", "hash", "base64", "uuid", "json_query", "math_eval", "text_stats":
		return dec("safe", "pure computation", fmt.Sprintf("Using %s", toolName))

	case "file_stat":
		path := getString(norm, "path")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "Path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		return dec("safe", "read-only", "Reading file metadata")

	case "env":
		return dec("safe", "read-only", "Reading environment variables")

	case "dns_lookup", "port_check", "service_list", "registry_read":
		return dec("safe", "read-only", fmt.Sprintf("Using %s", toolName))

	case "service_control":
		return dec("confirm", "changes a system service", fmt.Sprintf("%s service", getString(norm, "action")),
			fmt.Sprintf("name: %s", getString(norm, "name")))

	case "app_launch":
		return dec("confirm", "launches an application", "Launching application",
			fmt.Sprintf("path: %s", getString(norm, "path")))

	case "download":
		return dec("confirm", "writes a file from the network", "Downloading to a file",
			fmt.Sprintf("url: %s", getString(norm, "url")))

	case "open_path":
		return dec("confirm", "launches an external program", "Opening with default app",
			fmt.Sprintf("path: %s", getString(norm, "path")))

	case "recycle":
		path := getString(norm, "path")
		if isDangerousPath(path) {
			return dec("blocked", "dangerous path", "Path is outside allowed scope",
				"UNC path, device path, or traversal detected")
		}
		if !isPathWithinWorkspace(path, workspace) {
			return dec("blocked", "recycles outside workspace", "Recycling outside workspace is blocked")
		}
		return dec("confirm", "destructive file operation", "Sending to Recycle Bin")

	case "notify":
		return dec("safe", "shows a notification", "Sending desktop notification")

	case "memory":
		return dec("safe", "local notes store",
			fmt.Sprintf("memory %s", getString(norm, "action")))

	case "http_request":
		method := strings.ToUpper(getString(norm, "method"))
		if method == "" {
			method = "GET"
		}
		url := getString(norm, "url")
		switch method {
		case "GET", "HEAD", "OPTIONS":
			return dec("safe", "read-only request", fmt.Sprintf("%s %s", method, url))
		default:
			return dec("confirm", "sends a write request", fmt.Sprintf("%s %s", method, url),
				"can create or modify remote state")
		}

	case "browser_click":
		return dec("confirm", "interacts with a web page", "Clicking page element",
			fmt.Sprintf("selector: %s", getString(norm, "selector")))

	case "browser_fill":
		return dec("confirm", "interacts with a web page", "Filling page field",
			fmt.Sprintf("selector: %s", getString(norm, "selector")))

	default:
		return dec("confirm", "unknown tool", fmt.Sprintf("Unknown tool: %s", toolName),
			"no safety policy defined — requires confirmation")
	}
}

func capitalize(s string) string {
	if s == "" {
		return ""
	}
	return strings.ToUpper(s[:1]) + s[1:]
}
